Add `code_ircode`

tkf opened the pull request:
When working on code like #45305, it would be helpful if we can easily get "real" `IRCode` examples. It's also useful for writing tests. So, I wonder if we can have a `code_typed`-like API for `IRCode`?

tkf committed:
Add typeinf_ircode

gbaraldi commented:
This looks very useful. Your ShowCode.jl package has been very nice for working at the IRCode level

ianatol commented on base/compiler/typeinfer.jl:
```diff
@@ -918,6 +918,33 @@ function typeinf_code(interp::AbstractInterpreter, method::Method, @nospecialize
+    end
+    opt_params = OptimizationParams(interp)
+    opt = OptimizationState(frame, opt_params, interp)
+    ir = run_passes(opt.src, opt, result)
```
Looks good overall! Would adding an option to toggle optimization of this IR be useful?

aviatesk replied:
Yeah, it sounds reasonable to have an option to return `IRCode` that hasn't been optimized yet (i.e. that just created from `convert_to_ircode`).

tkf replied:
I added the `optimize` flag d5a9dc2b5a8537c34e328b80c9b73d55aa076480

```julia
julia> Base.code_ircode(; optimize = false) do f
           identity(f())
           identity(f())
           f()
       end |> only |> first
2 1 ─ %1 = (_2)()::Any                                                                             │
  │        Main.identity(%1)::Any                                                                  │
3 │   %3 = (_2)()::Any                                                                             │
  │        Main.identity(%3)::Any                                                                  │
4 │   %5 = (_2)()::Any                                                                             │
  └──      return %5                                                                               │


julia> Base.code_ircode() do f
           identity(f())
           identity(f())
           f()
       end |> only |> first
2 1 ─      (_2)()::Any                                                                             │
3 │        (_2)()::Any                                                                             │
4 │   %3 = (_2)()::Any                                                                             │
  └──      return %3                                                                               │
```

I created `run_minimal_passes` to run `convert_to_ircode`, `slot2reg`, and `compact!`: https://github.com/JuliaLang/julia/blob/d5a9dc2b5a8537c34e328b80c9b73d55aa076480/base/compiler/optimize.jl#L548

Since most of the transformations on IRCode do not handle slots, I think it's better to run `slot2reg`. But maybe we should skip `compact!` when `optimize = false`? What do you think?

aviatesk replied:
I agree with running `compact!` as I found it more useful to run `compact!` when writing a similar tooling for EA.jl (https://github.com/JuliaLang/julia/blob/d6a4417af880d140e4e3702550906c8fc25bdd50/test/compiler/EscapeAnalysis/EAUtils.jl#L223) especially because we will see some "unfixed" states  that make the testing a bit difficult without `compact!`.

tkf replied:
Yeah, that's why I wanted to include `compact!` even with `optimize = false`. A non-`compact!`-ed IR is a bit tricky to work with and also the printing is misleading sometimes.

`code_ircode` is a "hidden" feature for interactive use so I don't think we need to be picky about getting it "right" at the first shot. So I suggest keeping it as-is (i.e., include `compact!`) for now unless someone with strong opinions comes before merging this.

aviatesk commented on base/compiler/typeinfer.jl:
```diff
@@ -918,6 +918,33 @@ function typeinf_code(interp::AbstractInterpreter, method::Method, @nospecialize
+
+Infer a `method` and return an optimized `IRCode` with inferred `returntype` on success.
+"""
+function typeinf_ircode(interp::AbstractInterpreter, method::Method, @nospecialize(atype), sparams::SimpleVector)
+    mi = specialize_method(method, atype, sparams)::MethodInstance
+    ccall(:jl_typeinf_begin, Cvoid, ())
+    result = InferenceResult(mi)
+    frame = InferenceState(result, :no, interp)
+    if !(frame !== nothing && typeinf(interp, frame) && frame.inferred)
+        ccall(:jl_typeinf_end, Cvoid, ())
+        return nothing, Any
+    end
+    opt_params = OptimizationParams(interp)
+    opt = OptimizationState(frame, opt_params, interp)
+    ir = run_passes(opt.src, opt, result)
+    ccall(:jl_typeinf_end, Cvoid, ())
+    rt = widenconst(ignorelimited(result.result))
+    return ir, rt
+end
```
```suggestion
function typeinf_ircode(interp::AbstractInterpreter, method::Method, @nospecialize(atype), sparams::SimpleVector)
    frame = typeinf_frame(interp, method, atype, sparams, false)
    frame === nothing && return nothing, Any
    opt_params = OptimizationParams(interp)
    opt = OptimizationState(frame, opt_params, interp)
    ir = run_passes(opt.src, opt, result)
    rt = widenconst(ignorelimited(result.result))
    return ir, rt
end
```

tkf replied:
I just want to double-check, but is it OK to run the optimizer outside the typeinf lock?

cc @vtjnash @pchintalapudi

tkf committed:
Use `typeinf_frame`

Co-authored-by: Shuhei Kadowaki <40514306+aviatesk@users.noreply.github.com>

tkf committed:
Add `optimize` option

tkf committed:
Tweak docstring

tkf committed:
Merge branch 'master' into typeinf-ircode

tkf committed:
Use typeinf lock just in case

To match `typeinf_ircode` with how typeinf lock is used ATM (i.e.,
optimizer is run inside the lock), we can manually lock it because the
lock is re-entrant.

aviatesk commented:
A very exciting showcase of this PR!: https://nbviewer.org/gist/tkf/d4734be24d2694a3afd669f8f50e6b0f/00_notebook.ipynb

tkf commented:
I guess you mean #45305 :)

Also, FYI, I'll probably tweak the API a little bit. So, the exact API may be a bit different by the time we merge this. But it'll have a similar flavor.

aviatesk reviewed (approved):
Yep, let's go ahead!

ianatol commented:
Yup, looks good!

tkf committed:
Doctest code_ircode

tkf commented on base/compiler/optimize.jl:
```diff
@@ -534,6 +534,26 @@ function ipo_escape_cache(mi_cache::MICache) where MICache
+"""
+function run_minimal_passes(ci::CodeInfo, sv::OptimizationState, caller::InferenceResult)
+    @timeit "convert"   ir = convert_to_ircode(ci, sv)
+    @timeit "slot2reg"  ir = slot2reg(ir, ci, sv)
```
I think it would be very rare to deal with `IRCode` without slog2reg (unless you are changing slog2reg itself). For one thing, `verify_ir` considers `IRCode` with slots invalid:

https://github.com/JuliaLang/julia/blob/2159bfba2f551bb22a92c6895ff3e46aa6dc5421/base/compiler/ssair/verify.jl#L47-L48

So, I think running slot2reg with `optimize = false` makes sense, if we treat it as "least optimized typical `IRCode` that compiler passes deal with." But running slot2reg manually is not super hard. So I don't have super strong opinion about it.

@aviatesk @ianatol Any thoughts?

tkf replied:
I guess we can also pass numeric `optimize_level`

```julia
function run_passes(ci::CodeInfo, sv::OptimizationState, caller::InferenceResult, optimize_level = typeinf(Int))
    stage = 1
    isdone() = optimize_level < (stage += 1)
    @timeit "convert"   ir = convert_to_ircode(ci, sv)
    isdone() && return ir
    @timeit "slot2reg"  ir = slot2reg(ir, ci, sv)
    isdone() && return ir
    # TODO: Domsorting can produce an updated domtree - no need to recompute here
    @timeit "compact 1" ir = compact!(ir)
    isdone() && return ir
    @timeit "Inlining"  ir = ssa_inlining_pass!(ir, ir.linetable, sv.inlining, ci.propagate_inbounds)
    isdone() && return ir
    ...
```

aviatesk replied:
I prefer running `slot2reg` as we often work on post-`slot2reg` `IRCode` when developing Julia-level optimizer. I also think the other "pre-inlining optimizations (mostly DCE)" happening within `convert_to_ircode` and `type_annotate!` is also significant anyway.

And the `optimize_level` idea sounds very useful!

tkf replied:
I added `optimize_level` 04797d59d3af59bb7db05a600b61151059fcaad0

```julia
julia> function demo(f)
           y = f()
           y isa Any ? nothing : f()
       end;

julia> Base.code_ircode(demo; optimize_level = 1) |> only |> first
2 1 ─      (_3 = (_2)())::#UNDEF                                                                   │
3 │   %2 = (_3 isa Main.Any)::#UNDEF                                                               │
  └──      goto #5 if not %2::#UNDEF                                                               │
  2 ─      return Main.nothing::#UNDEF                                                             │
  3 ─      Core.Const(:((_2)()))::#UNDEF                                                           │
  │        Core.Const(:(return %5))::#UNDEF                                                        │
  └──      unreachable::#UNDEF                                                                     │


julia> Base.code_ircode(demo; optimize_level = 2) |> only |> first
2 1 ─ %1 = (_2)()::Any                                                                             │
3 │   %2 = (%1 isa Main.Any)::Core.Const(true)                                                     │
  └──      goto #3 if not %2                                                                       │
  2 ─      return Main.nothing                                                                     │
  3 ─      Core.Const(:((_2)()))::Union{}                                                          │
  │        Core.Const(:(return %5))::Union{}                                                       │
  └──      unreachable                                                                             │


julia> Base.code_ircode(demo) |> only |> first
2 1 ─     (_2)()::Any                                                                              │
3 └──     goto #3 if not true                                                                      │
  2 ─     return Main.nothing                                                                      │
  3 ─     unreachable                                                                              │

```

tkf committed:
Add `optimize_level`

aviatesk commented on base/compiler/optimize.jl:
```diff
@@ -534,18 +534,30 @@ function ipo_escape_cache(mi_cache::MICache) where MICache
+    macrocall.args[2] = __source__  # `@timeit` may want to use it
+    quote
+        $macrocall
+        $optimize_level < ($stage += 1) && return $(esc(:ir))
```
How about this? So that we can specify the optimizer phase more directly:
```suggestion
        if $optimize_level isa String
            $optimize_level == $(esc(name)) && return $(esc(:ir))
        else
            $optimize_level < ($stage += 1) && return $(esc(:ir))
        end
```

tkf replied:
Good idea! I added it in ae3679d8114f47c9602ff7bad2107bb675f9f199

tkf committed:
Don't refer to the variable `ir` in `@pass`

tkf committed:
Support specifying pass by name

tkf committed:
Fix code style

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

aviatesk commented:
Just in case: @nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/45306#issuecomment-1132642484) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/70f4e65_vs_b81a20c/report.md).

tkf commented:
Good to go?

@vtjnash Do you want to review this?
